forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,169 @@
+                final boolean play = p.getController().chooseBinary(sa, prompt,
+                        PlayerController.BinaryChoiceType.CastOrHand);
+
+                final Zone origin = found.getZone();
+                if (play) {
+                    // get basic spells (no flashback, etc.)
+                    List<SpellAbility> sas = AbilityUtils.getBasicSpellsFromPlayEffect(found, p, CardStateName.Original);
+
+                    // filter land abilities due to MDFC or similar
+                    List <SpellAbility> landAbilities = Lists.newArrayList();
+                    for (SpellAbility s : sas) {
+                        if (s instanceof LandAbility) {
+                            landAbilities.add(s);
+                        }
+                    }
+                    sas.removeAll(landAbilities);
+
+                    if (!sas.isEmpty()) {
+                        SpellAbility tgtSA = p.getController().getAbilityToPlay(found, sas);
+
+                        // in case player canceled from choice dialog
+                        if (tgtSA != null) {
+                            tgtSA = tgtSA.copyWithNoManaCost();
+
+                            // 118.8c
+                            boolean optional = false;
+                            for (CostPart cost : tgtSA.getPayCosts().getCostParts()) {
+                                if ((cost instanceof CostDiscard || cost instanceof CostReveal)
+                                        && !cost.getType().equals("Card") && !cost.getType().equals("Random")) {
+                                    optional = true;
+                                    break;
+                                }
+                            }
+                            if (!optional) {
+                                tgtSA.getPayCosts().setMandatory(true);
+                            }
+
+                            if (tgtSA.usesTargeting() && !optional) {
+                                tgtSA.getTargetRestrictions().setMandatory(true);
+                            }
+
+                            tgtSA.setSVar("IsCastFromPlayEffect", "True");
+
+                            if (p.getController().playSaFromPlayEffect(tgtSA)) {
+                                final Card played = tgtSA.getHostCard();
+                                // remember successfully played here if ever needed
+                                final Zone zone = game.getCardState(played).getZone();
+                                if (!origin.equals(zone)) {
+                                    CardZoneTable trigList = new CardZoneTable();
+                                    trigList.put(origin.getZoneType(), zone.getZoneType(), game.getCardState(found));
+                                    trigList.triggerChangesZoneAll(game, sa);
+                                }
+                            }
+                        }
+                    }
+                } else {
```

Comment:
in case the dialog is canceled, the card should be put in hand too?

Reply:
Think I've got this sorted